fix: add security gate for external contributor PRs

volnei opened the pull request:
## What does this PR do?

Adds a security gate to prevent the `pull_request_target` + `dangerous-git-checkout` vulnerability where external contributors could execute arbitrary code with access to repository secrets.

**Background:** A malicious PR (#26291) demonstrated this vulnerability by modifying `package.json` to exfiltrate data when `yarn lint` ran. The attacker's OAST callback succeeded (HTTP 200), proving arbitrary code execution with secrets was possible.

**Changes:**
- Adds a `trust-check` job that gates all CI jobs executing PR code with secrets
- Trusted contributors (OWNER, MEMBER, COLLABORATOR) can run CI immediately
- External contributors require approval from a core team member (write/maintain/admin permission)
- Approval must be for the current commit SHA (stale approvals are ignored, works with "dismiss stale reviews on new pushes")
- Adds `pull_request_review` trigger so CI starts when approval is given
- Updates `required` job to fail with clear message for untrusted PRs
- Handles `workflow_dispatch` events as trusted (only users with write access can trigger manually)
- Non-approval review events (comments, changes_requested) go into a separate `-noop` concurrency group to avoid canceling in-progress CI

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a documentation change. N/A - workflow change only.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works. N/A - GitHub Actions workflow changes cannot be unit tested.

## How should this be tested?

1. **Test with trusted contributor:** Create a PR from a repo member - CI should run immediately
2. **Test with external contributor:** Have someone outside the org create a PR - CI should NOT run, `required` check should fail with message about needing approval
3. **Test approval flow:** Have a core team member approve the external PR - CI should start running
4. **Test stale approval:** After approval, push a new commit - CI should stop (approval was for old SHA), require re-approval
5. **Test manual trigger:** Use "Run workflow" button - should work without PR context
6. **Test comment review:** Leave a comment on a PR - should NOT cancel in-progress CI (goes to `-noop` group)

## Checklist

- [x] I have read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have checked if my changes generate no new warnings

## Human Review Checklist

- [x] Verify the `trust-check` job correctly handles `workflow_dispatch` (no PR context) - Added handling to treat as trusted
- [x] Verify the `pull_request_review` trigger doesn't cause issues with concurrency groups - Fixed with `-noop` group for non-approval reviews
- [x] Verify the `required` job handles skipped `trust-check` vs `is-trusted: false` correctly - Fixed with `&& needs.trust-check.result == 'success'`
- [ ] Consider edge cases: approval + new commit race condition, reviewer permission check failures
- [ ] Confirm `getCollaboratorPermissionLevel` API call has sufficient permissions

---

**Link to Devin run:** https://app.devin.ai/sessions/52cdd1ebd9c64aa8be53d70171ddd6e1
**Requested by:** Volnei Munhoz (@volnei)

devin-ai-integration[bot] committed:
fix: add security gate for external contributor PRs

This change adds a trust-check job that gates CI for external contributors:

- Trusted contributors (OWNER, MEMBER, COLLABORATOR) can run CI immediately
- External contributors require approval from a core team member (write access or higher)
- Approval must be for the current commit SHA (stale approvals are ignored)
- Added pull_request_review trigger so CI runs when approval is given
- The required job now fails with a clear message for untrusted PRs

This prevents the pull_request_target + dangerous-git-checkout vulnerability
where external contributors could execute arbitrary code with access to secrets.

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

keithwillcode commented on .github/workflows/pr.yml:
```diff
@@ -18,8 +21,118 @@ concurrency:
+    name: Trust Check
+    runs-on: blacksmith-2vcpu-ubuntu-2404
+    # Skip if this is a non-approval review event (e.g., comment or changes_requested)
+    if: github.event_name != 'pull_request_review' || github.event.review.state == 'approved'
```
DevinAI, why are we skipping this? The trust check should always run, no?

keithwillcode replied:
DevinAI, especially since you didn't change pull_request_target to pull_request

devin-ai-integration[bot] committed:
fix: handle workflow_dispatch event in trust-check

Added handling for workflow_dispatch events which don't have PR context.
Manual triggers are now assumed trusted since only users with write access
can trigger workflow_dispatch.

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

pedroccastro commented on .github/workflows/pr.yml:
```diff
@@ -279,6 +398,12 @@ jobs:
+          echo "::error::This PR is from an external contributor and requires approval from a core team member before CI can run."
+          echo "A maintainer with write access must approve this PR to trigger CI checks."
+          exit 1
+        if: needs.trust-check.outputs.is-trusted != 'true'
```
If trust-check is skipped (e.g., a comment review event), is-trusted will be undefined. The current check `needs.trust-check.outputs.is-trusted != 'true'` will evaluate to true and fail the job unnecessarily

Suggestion:

`if: needs.trust-check.outputs.is-trusted != 'true' && needs.trust-check.result == 'success'`

pedroccastro commented on .github/workflows/pr.yml:
With the current concurrency config, a comment review triggers a new workflow run in the same concurrency group, which cancels any in-progress CI

Suggestion:

```
concurrency:
  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}${{ github.event_name == 'pull_request_review' && github.event.review.state != 'approved' && '-noop' || '' }}
  cancel-in-progress: ${{ github.event_name != 'pull_request_review' || github.event.review.state == 'approved' }}
```

This would put non-approval reviews in a separate -noop group so they don't interfere with running CI

devin-ai-integration[bot] committed:
fix: handle skipped trust-check and prevent non-approval reviews from canceling CI

Co-Authored-By: Volnei Munhoz <volnei.munhoz@gmail.com>

pedroccastro committed:
Merge branch 'main' into devin/secure-pr-workflow-1767046413

keithwillcode committed:
Merge branch 'main' into devin/secure-pr-workflow-1767046413
